Add test for challenger and warning

NimaSarajpoor opened the pull request:
This PR is to add test for the modules in `/sdp/`

NimaSarajpoor commented on test.py:
```diff
@@ -15,17 +15,21 @@ def func_exists(mod_path, func_name):
             module_content = file.read()
-    except FileNotFoundError:
+    except FileNotFoundError as e:
+        warnings.warn(f"SKIPPED: {mod_path},{func_name}: \n{e}")
```
According to the comment in the next line of code, maybe:

```
msg = f"SKIPPED: Module file not found. \n{e}"
warnings.warn(msg)
```

NimaSarajpoor replied:
@seanlaw 
This and some other comments are outdated since this part is now moved to a different file. If you do not mind, I can resolve it here for now, and mention it later. Not a major concern IMO.

seanlaw replied:
Sounds good. I saw the comment but also saw that it was outdated so I ignored it

NimaSarajpoor commented on test.py:
```diff
@@ -15,17 +15,21 @@ def func_exists(mod_path, func_name):
         tree = ast.parse(module_content)
-    except SyntaxError:
+    except SyntaxError as e:
+        warnings.warn(f"SKIPPED: {mod_path},{func_name}: \n{e}")
```
According to the comment in the next line of code, maybe:

msg = f"SKIPPED: Syntax error in module. \n{e}"
warnings.warn(msg)

NimaSarajpoor commented on test.py:
```diff
@@ -15,17 +15,21 @@ def func_exists(mod_path, func_name):
         if isinstance(node, ast.FunctionDef) and node.name == func_name:
             return True
+    e = f"Function {func_name} not found in {mod_path}"
+    warnings.warn(f"SKIPPED: {mod_path},{func_name}: \n{e}")
```
Suggestion:

```
msg = "SKIPPED : function not found in module."
warnings.warn(msg)
```

NimaSarajpoor reviewed (commented):
@seanlaw 
Once you find some time, can you please review my comments and add yours?

NimaSarajpoor commented on test_challenger.py:
```diff
@@ -0,0 +1,29 @@
+import pytest
+
+from numpy import testing as npt
+from sdp import challenger_sdp
```
[As suggested](https://github.com/stumpy-dev/sliding_dot_product/pull/13#discussion_r1976654151), 
will update this by testing modules that have both `sliding_dot_product ` & `setup`, for range 3 to 13 (inclusive).

NimaSarajpoor commented:
**[Update]**
Moved some functions to `utils.py`, and the unit testing now covers all modules that have `setup` & `sliding_dot_product`.

seanlaw commented on test.py:
```diff
@@ -10,53 +10,7 @@
-
-    return mods
-
+from utils import import_sdp_mods
```
1. Let's just `import utils` as writing `import import_sdp_mods` looks bad
2. Now that we have unit tests, let's rename this `test.py` file to `timing.py` (and reserve "test" related things for unit testing)

NimaSarajpoor replied:
Yeah... was thinking about `test_performance.py` but wanted to avoid `test `... `timing.py` is a good name! 👍

seanlaw commented:
For testing, it would be good to also test non-power-of-2 lengths. So, say, p-1 and p+1 or large-ish prime numbers

I feel uneasy only testing perfect powers of two

NimaSarajpoor commented on test_modules.py:
```diff
@@ -0,0 +1,38 @@
+import pytest
+
+from numpy import testing as npt
+from sdp import challenger_sdp
```
Correct. and I noticed some other issues as well. Will add `flake8` (and `black`) to Github Actions.

NimaSarajpoor commented:
@seanlaw 
> I feel uneasy only testing perfect powers of two

Right! I will revise the unit testing.

NimaSarajpoor commented:
@seanlaw 

> For testing, it would be good to also test non-power-of-2 lengths. So, say, p-1 and p+1 or large-ish prime numbers.
> I feel uneasy only testing perfect powers of two

Added test and got error locally in one of sdp's module. Will add GitHub Actions to expose the error here first, and then work on fixing it here (or in another PR)

seanlaw commented:
Let's try to fix it here

NimaSarajpoor commented:
**[Update]**
I can now see the [assertion error in Github Actions on my fork](https://github.com/NimaSarajpoor/sliding_dot_product/actions/runs/13691554121/job/38285576232). I tried to see if I can somehow trigger it here before merging it to main but couldn't. 

Going to fix the error.

NimaSarajpoor commented:
@seanlaw 
The bug is fixed and the Github Actions' workflow run is now successful in my fork's branch:
https://github.com/NimaSarajpoor/sliding_dot_product/actions/runs/13692087301

**A few notes regarding the `gitHub-actions.yml` file:**
(1) it does unit testing on different platforms for python >=3.10 to satisfy [pyfftw's requirement](https://github.com/pyFFTW/pyFFTW/blob/9bbc6da44f63b74c5508915a0c84aeda01656f73/pyproject.toml#L23)
(2) it checks black and flake8 but no isort
(3) it does unit test but no coverage
(4) it is created by using [STUMPY's gitHub-actions file](https://github.com/TDAmeritrade/stumpy/blob/main/.github/workflows/github-actions.yml)

NimaSarajpoor commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,37 @@
+name: Tests
+on:
+  push:
+    branches: '**'
```
Will replace `**` with `main`

seanlaw replied:
Don't forget this

NimaSarajpoor commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,37 @@
+        run: python -m pip install --upgrade -r requirements.txt
+        shell: bash
+      - name: Install other dependencies
+        run: python -m pip install --upgrade pytest black flake8
```
No minimum version is specified for these libraries here.

seanlaw replied:
Yes

seanlaw replied:
I don't even think we need to test so many versions. Instead, is it possible to simply choose a reasonably stable Python version?

NimaSarajpoor replied:
>  Instead, is it possible to simply choose a reasonably stable Python version?

How about choosing one-before-last [major] version?

seanlaw replied:
You might hate this but I wrote an ugly `curl` function that picks the "safe" version of Python using some xpath magic:

```
#!/bin/bash

get_safe_python_version()
{
    SAFE_PYTHON=curl --location https://devguide.python.org/versions | xmllint --html --xpath '//section[@id="supported-versions"]//table/tbody/tr[count(//section[@id="supported-versions"]//table/tbody/tr[td[.="security"]]/preceding-sibling::*)]/td[1]/p/text()' - 2>/dev/null
}

get_safe_python_version
echo $SAFE_PYTHON
```

Basically, it scrapes `https://devguide.python.org/versions`, looks for the latest "security" version (yellow), and takes the Python version that is one less than the latest security version.

NimaSarajpoor replied:
@seanlaw 
Cool! IIRC, I think I saw, somewhere in STUMPY, that you used `pandas` to read table and it was really clean. So, can this work?

```
import pandas as pd
from packaging.version import Version
df = pd.read_html("https://devguide.python.org/versions/")[0]
out = df[df.Status == 'security'].Branch.tolist()
out = sorted(out, key=lambda x: Version(x), reverse=True)
print(out[1])  # 3.10
```

seanlaw replied:
Yes but now this adds a new dependency that requires you to install `pandas` in the Github workflow environment just so you can do this. Whereas the one-liner "just works" out of the box.

seanlaw replied:
> nvm. Because now I think I need a way to return that value and assign it to a variable in shell.

Exactly! This (ugly) one liner will get you there immediately

NimaSarajpoor replied:
Yeah...And the "trade-off" you mentioned makes sense here as we most likely do not need to revisit it in the future.

NimaSarajpoor replied:
Thanks for sharing the `curl`! Going to revise Github Actions.

seanlaw reviewed (commented):
@NimaSarajpoor I left some quick comments

NimaSarajpoor commented on sdp/numpy_fft_sdp.py:
```diff
@@ -14,7 +14,8 @@ def sliding_dot_product(Q, T, order="F"):
     tmp[0, m:] = 0.0
-    tmp[1, :] = T
+    tmp[1, :n] = T
+    tmp[1, n:] = 0.0
```
I need to check again and make sure a test is failing for `tmp[1, :]=T`. Because the root cause of the observed failure was resolved when I pass `n=shape` to the [irfft function](https://numpy.org/doc/stable/reference/generated/numpy.fft.irfft.html) but, before such change, I noticed `tmp[1, :] = T` and fixed it. 

> I need to check again and make sure a test is failing for `tmp[1, :]=T`

I will work on this, and will consider your `large-ish prime number` comment as well to see if that is needed. As long as we have a prime number that is considered as "bad factor for effiecent FFT" , then `shape != n`, and we will probably see failure for `tmp[1, :]=T`. I will check and report back here.

seanlaw replied:
Ahh, I missed that `n=shape` is passed to `irfft`. This might cause a decrease in performance (and I should also incorporate this into the `r2c`/`c2r`). 

It's possible that `n = shape` is not needed in `irfft`. Thanks for checking it

NimaSarajpoor replied:
This is what I noticed after reading the doc: 

For T with length `n`, the length of `RFFT(T)` is `n / 2` if n is even, and  if `n` is odd, the length is `(n + 1) / 2` (See [doc](https://numpy.org/doc/2.2/reference/generated/numpy.fft.rfft.html)). So, I can say the length of RFFT is `n // 2 + 1` for any `n`.  So, when we do `IRFFT(arr)`, IRFFT has no idea about `n` (the length of original array in time domain). According to numpy's doc, when the value for param `n` is not provided in IRFFT, it assumes the length of output is `2 * (m - 1)`, which means that it assumes the length of array in time domain is even. So, that is why we were good previously. But the moment we test it for array with odd length, it failed.

NimaSarajpoor replied:
@seanlaw 
I think so. The `c2r` code seems to make the same assumption

https://github.com/scipy/scipy/blob/0f1fd4a7268b813fa2b844ca6038e4dfdf90084a/scipy/fft/_pocketfft/basic.py#L87-L88

Also, once this PR is merged, it should be able to catch it.

seanlaw replied:
Hmm, it looks like I `c2r` automatically handles this for me (i.e., adding `n=shape` has no effect and excluding it still gives the correct answer for the `1 + 2 ** x` case). This is good because it means that I won't have to re-run my timings.

seanlaw replied:
> I think so. The c2r code seems to make the same assumption

@NimaSarajpoor I can't quite tell if specifying the `n=shape` will make the `tmp` array shorter (and therefore make the `irfft` faster)?

NimaSarajpoor replied:
That was interesting, and it was tricky to find out why! 

The slice `m -1 : n` exists in output with correct values when `shape` > `n=len(T)`. However, if `shape` (the output of `next_best_len`) is exactly the same as `n`, we will face wrong outputs when `n` is odd. See example below:

``` 
import numpy as np
from scipy.fft import next_fast_len
from scipy.fft._pocketfft.basic import r2c, c2r


def sliding_dot_product(Q, T):
    n = len(T)
    m = len(Q)
    shape = next_fast_len(n)

    print('n: ', n)
    print('shape: ', shape)

    tmp = np.empty((2, shape))
    tmp[0, :m] = Q[::-1]
    tmp[0, m:] = 0.0
    tmp[1, :len(T)] = T
    tmp[1, len(T):] = 0
    fft_2d = r2c(True, tmp, axis=-1)

    return c2r(False, np.multiply(fft_2d[0], fft_2d[1]))[m - 1:n]


def naive_sliding_dot_product(Q, T):
    m = len(Q)
    l = T.shape[0] - m + 1
    out = np.empty(l)
    for i in range(l):
        out[i] = np.dot(Q, T[i : i + m])
    return out
    
 
n_T = 2 ** 3 + 1
n_Q = 5

T = np.random.rand(n_T)
Q = np.random.rand(n_Q)

ref = naive_sliding_dot_product(Q, T)
comp = sliding_dot_product(Q, T)
np.testing.assert_allclose(comp, ref)
```


And the assertion fails here. In this case, `n` has the "good" factors that are efficient for FFT computation and hence the output of `next_best_length(9)` is `9`. In this case, the size of `F` will be `9//2 + 1 = 5`, and the length of inverse FFT will be `2 * ( 5 - 1) = 8`, and hence it does not cover the slice `m - 1 : n`

------

Btw, there was a mistake in my text:

> For T with length n, **the length of RFFT(T) is n / 2 if n is even**, and if n is odd, the length is (n + 1) / 2 (See [doc](https://numpy.org/doc/2.2/reference/generated/numpy.fft.rfft.html)). So, I can say the length of RFFT is n // 2 + 1 when input array has length n

correction: **the length of RFFT(T) is `(n / 2) + 1` if n is even**,

seanlaw replied:
> And the assertion fails here. In this case, n has the "good" factors that are efficient for FFT computation and hence the output of next_best_length(9) is 9. In this case, the size of F will be 9//2 + 1 = 5, and the length of inverse FFT (without providing its length) will be 2 * ( 5 - 1) = 8, and hence it does not cover the slice m - 1 : n

@NimaSarajpoor Please make sure to add this test case to our unit tests. 

Perhaps we should test prime number inputs like 7, 11, 13, 17, 19, 23, and (to account for oaconvolve blocks/chunks) maybe a larger one like one of:

307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397

NimaSarajpoor replied:
@seanlaw 

> Please make sure to add this test case to our unit tests.

Sure. Will do!

> Perhaps we should test prime-only

Good idea! And I think you meant in addition to the test for "numbers that are complete power of two". Right? And also we should test it for... (see below)

> and this will also naturally cover the "odd" case?

To test that odd case, I think I need to test it with small / large numbers that contain only the factors 3 and 5 (I think numbers with these factors are computationally efficient from FFT's perspective). So, basically, any number in the form of  `(3^x) * (5^y)` should be a good cancidate... because it is odd AND its `next_fast_len` should be the same as the length of original array, and the test should fail if one forgets to pass `n=shape` when calling IRFFT. Need to check it to make sure of it.

NimaSarajpoor replied:
I was thinking more on "what" we want to test... I think we have five cases:

(Case1) 2^x
(Case2) (2^x) * (3^y) * (5^z), where y+z>=1

(Case3) (2^x) * (LargePrime^w)

(Case4) (3^y) * (5^z), where y+z>=1

(Case5) LargePrime^w

So, if I put the first case aside, the other four cases are basically the possible combination of even/odd AND bestshape is/isn't equal to len(T)

seanlaw replied:
Sounds good!

NimaSarajpoor replied:
**[Update]**
I improved our test cases in `test.py` based on our discussion here and they are passing locally. Once Github-Actions is fixed, the tests should pass on Github too. Then I will share some comments. So, I am going to resolve this conversation.

NimaSarajpoor commented on timing.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+source test.sh
```
I did not add execution permission. I added it to the file later.  So, I think I should be able to remove `source` and just do `./test.sh` now.

NimaSarajpoor replied:
Done

NimaSarajpoor commented:
@seanlaw 
> Was the bug related to adding zeros when the shape was not equal to the length of T?

Not the one that was observed. I need to check again and make sure a test is failing for `tmp[1, :]=T`. Because the root cause of the observed failure was resolved when I pass `n=shape` to the [irfft function](https://numpy.org/doc/stable/reference/generated/numpy.fft.irfft.html). Read more [here](https://github.com/stumpy-dev/sliding_dot_product/pull/14#discussion_r1983826060). 


> I see that pyFFTW is required but FFTW isn't required. How/when is FFTW installed?

I thought installing `pyFFTW` will take care of it. I checked the log in Github Actions but did not find FFTW. I think I am missing something here!

NimaSarajpoor commented:
@seanlaw 
I think it is time for me to get help!! 

Could you please take a look at [this Github-Actions error](https://github.com/NimaSarajpoor/sliding_dot_product/actions/runs/13751028882/job/38452017784). I am getting 
```
curl: (23) Failure writing output to destination
```

and cannot figure out why this happens (and how to solve it)

seanlaw commented:
Okay, I will take a look

NimaSarajpoor commented:
**Update regarding Github-Actions**
I tried to break down the process to catch the issue on GitHub Actions. I wrote output of curl command to `tmp.txt`. Then, in the next command I did: `cat tmp.txt | xmllint ..`, and then noticed [a new error message on GitHub Actions](https://github.com/NimaSarajpoor/sliding_dot_product/actions/runs/13752589464/job/38455434178):

> cat: write error: Broken pipe

Btw, I am now using a new branch to avoid adding more noise here

seanlaw commented:
> Okay, I will take a look

@NimaSarajpoor Please take a look at [this repo](https://github.com/seanlaw/safe_python). I believe that the issue is that `xmllint` is not installed by default on Linux (but is available on MacOS and Windows runners). If you take a look at my Github workflow, you'll see that I install `xmllint` when a `Linux` runner is detected. Let me know if that helps.

NimaSarajpoor commented:
@seanlaw 
THANK YOU!!!

NimaSarajpoor commented on test.py:
```diff
@@ -1,131 +1,175 @@
+                print(msg)
+                raise e
+
+    return
```
I did not add any test for `oaconvolve` because of two reasons:

(1) Having large `len(T)` alone does not guarantee that the overlap-add part of `oaconvolve` is going to be triggered as that depends on both `len(T)` and `len(Q)`. 

(2) I am now testing for a wide range of len(Q), from 2 to len(T). So, I think overlap-add part is going to be triggered at some point.

NimaSarajpoor replied:
The [module sdp/scipy_oaconvolve_sdp.py](https://github.com/NimaSarajpoor/sliding_dot_product/blob/84837766959e989bd0589fc221be858d37b58153/sdp/scipy_oaconvolve_sdp.py#L9-L10) uses the oaconvolve function from scipy and I do not think we can "easily" figure out whether it, internally, fallbacks to fftconvolve or not unless we go through that source code and check the fallback function  Still, your concern should be addressed:

> Well, you can be 100% sure of this if you check test coverage...

 Because, if we later add something like PR #17, then it is important to know if all of its parts are tested.

NimaSarajpoor replied:
My mistake was that I saw the current modules... and said: "okay...no coverage is needed". That was a very bad mindset. I should have thought about keeping ourselves safe for the long term when other new modules are added.

seanlaw replied:
It's okay. You're not wrong. You made a rational decision at the time. Now, the circumstances are changing and so we should update our perspective accordingly. :)

NimaSarajpoor commented on sdp/numpy_fft_sdp.py:
```diff
@@ -14,7 +14,8 @@ def sliding_dot_product(Q, T, order="F"):
     fft_2d = np.fft.rfft(tmp, axis=-1)
 
-    return np.fft.irfft(np.multiply(fft_2d[0], fft_2d[1]))[m - 1 : n]
+    return np.fft.irfft(np.multiply(fft_2d[0], fft_2d[1]), n=shape)[m - 1 : n]
```
Just to recap, here we've made two changes in this file:

(1) fixing `tmp[1, :]=T`
(2) passing `n=shape` when calling irfft. 

With new tests, I can see that:
* without (1) & (2): `test_sdp_case2`, `test_sdp_case3`, and `test_sdp_case4` all fail.
* with only (1): `test_sdp_case2` fails
* with only (2): `test_sdp_case3` and `test_sdp_case4` fail.

NimaSarajpoor reviewed (commented):
@seanlaw 
Thanks again for your support on handling Github-Actions! I've left some comments. Can you please check and let me know if you have any opinion/ concern?

seanlaw commented:
> Btw, I am curious on how you detected it...

So, I simply followed a similar logic to you if you see the order of my changing workflows (i.e., new commits):

https://github.com/seanlaw/safe_python/actions

[First](https://github.com/seanlaw/safe_python/actions/runs/13755627532), I received a message about permissions so I changed the permissions.

[Next](https://github.com/seanlaw/safe_python/actions/runs/13755645709/job/38462560841), I got the dreaded `curl: (23) Failure writing output to destination` error. However, what I noticed was that `macos` and `windows` were both passing. Additionally, they were successfully printing out `3.10`. This gave me some idea that it may be Linux specific.

[Then](https://github.com/seanlaw/safe_python/actions/runs/13755714032/job/38462738626), I noticed that the Linux error said `./python_version.sh: line 16: xmllint: command not found`. Again, this error did not exist for `macos` and `windows` as they were successfully printing `3.10`. Certainly, this is a weird one that took me a few minutes to find and isolate.

seanlaw commented:
> @seanlaw Thanks again for your support on handling Github-Actions! I've left some comments. Can you please check and let me know if you have any opinion/ concern?

Let me take a look.

seanlaw reviewed (commented):
There seems to be a lot of repeated code in case 1-4. Namely,

```
    shape = next_fast_len(n_T)
    if shape <= n_T:
        msg = (
            "In this test, n_T value should be set to an odd value "
            + "such that its next_fast_len must be larger"
        )
        raise ValueError(msg)

    modules = utils.import_sdp_mods()
    for mod in modules:
        for n_Q in range(2, n_T + 1):
            try:
                Q = np.random.rand(n_Q)
                T = np.random.rand(n_T)

                ref = naive_sliding_dot_product(Q, T)
                comp = mod.sliding_dot_product(Q, T)
                npt.assert_allclose(comp, ref)

            except Exception as e:
                msg = f"Error in {mod.__name__}, with n_Q={n_Q} and n_T={n_T}"
                print(msg)
                raise e
```

1. Does it make sense to do `for n_Q in range(2, n_T + 1):`? Is it fast? How long does it take?
2. It feels like we should just parameterize the inputs and combine case 1-4 into something like:

```
test_n_T = [
    (2 * (3**2) * (5**3)),  # = 2250, Even `len(T)` and `len(T) == next_fast_len(len(T))`
    ((3**2) * (5**3)),  # = 1125, Odd `len(T)` and `len(T) == next_fast_len(len(T))`
    (2 * 7 * 11 * 13),  # = 2002, Even `len(T)` and `len(T) < next_fast_len(len(T))`
    (7 * 11 * 13)  # = 1001, Odd `len(T)` and `len(T) < next_fast_len(len(T))`
]

@pytest.mark.parametrize("n_T", test_n_T)
def test_sdp(n_T):
     ...
```

NimaSarajpoor commented on test.py:
```diff
@@ -1,131 +1,175 @@
+
+    n_T = 2 * (3**2) * (5**3)
+    shape = next_fast_len(n_T)
+    if shape != n_T:
```
Sounds much better, and cleaner!

NimaSarajpoor commented on test.py:
```diff
@@ -1,131 +1,175 @@
+    # the len(T) but longer. To this end, we choose
+    # factors 2, 7, 11, 13.
+
+    n_T = 2 * 7 * 11 * 13
```
Got it. Need to work on providing clear explanation.

NimaSarajpoor commented:
@seanlaw 
>  Is it fast? How long does it take?

All tests (i.e. all cases, all modules, and for all that `n_Q`) take `<1 min` per platform. This can increase when we add coverage (and we also turn off njit there. So, need to check how much that adds to the run time). 


> Does it make sense to do `for n_Q in range(2, n_T + 1):`?

I think we should be fine by just checking `n_Q =  n_T // 2` and the edge case `n_Q = n_T` (and maybe `n_Q = n_T - 1` ??). we are not checking the running time in test though. So, if a new module is  introduced that takes a lot of time in some cases, it has more chance to increase the running time when we do it for wide range of `n_Q`. So, although the running time is low, I think maybe I should test it for a few values of `n_Q `. 

What do you suggest?

----

> It feels like we should just parameterize the inputs and combine case 1-4 into something like:

This is much better. My only concern is that how we can check to make sure all combinations of `n_T % 2 == 0 (!= 0)` and `shape == n_T (!= n_T)` are covered. I am slightly inclined to go with the parametrized version because it hugely reduces the code and make it MUCH EASIER to check it later. But I wanted to highlight my concern as well. 

What do you think?

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,59 @@
+        uses: actions/checkout@v4
+      
+      - name: Set environment variables
+        run:  echo "PYTHON_VERSION=$(./python_version.sh)" >> $GITHUB_ENV
```
I question whether you need to have `pythn_version.sh` and, instead, do:

```
echo "PYTHON_VERSION=$(curl --location https://devguide.python.org/versions | xmllint --html --xpath '//section[@id="supported-versions"]//table/tbody/tr[count(//section[@id="supported-versions"]//table/tbody/tr[td[.="security"]]/preceding-sibling::*)]/td[1]/p/text()' - 2> /dev/null)" >> $GITHUB_ENV
```

NimaSarajpoor replied:
Not really. I wanted to avoid spreading the "ugliness" of curl function. But I can see there are two costs associated with it: (i) adding an extra file, and (ii) forcing a reader of Github-Action workflow to look away to see what `pythn_version.sh` does.

seanlaw replied:
Exactly!

seanlaw commented on test.py:
```diff
@@ -1,131 +1,123 @@
+# 4. len(T) is composed of a subset of {7, 11, 13, ...}
+# 5. len(T) is power of 2
+
+# In some cases, the prime factors are powered to a certain degree
```
"are raised to a power of a certain degree"? I can understand your meaning but "powered" is somewhat overloaded

NimaSarajpoor replied:
Right! That was not a good passive sentence 😅 I think what you suggested, i.e. `raised`, is the common term.

NimaSarajpoor commented:
This PR looks much better now! Thank you @seanlaw! I am going to think & work on three things:

(1) Is it possible to show GitHub Actions right here in this PR before merge? Then I do not need to check my fork every time.

(2) Fix coverage. Currently, it shows:
```
Name                          Stmts   Miss  Cover   Missing
-----------------------------------------------------------
sdp/challenger_sdp.py            10      1    90%   5
sdp/njit_sdp.py                  18      2    89%   20-21
sdp/numpy_fft_sdp.py             15      1    93%   6
sdp/pyfftw_sdp.py                28      2    93%   46-47
sdp/scipy_oaconvolve_sdp.py       6      1    83%   6
test.py                          57      8    86%   95-98, 122-125
-----------------------------------------------------------
TOTAL                           134     15    89%
```

(3) "Install FFTW (?)" . if needed, then I should try and see how to install it on GitHub Actions (via Conda? or ...)

NimaSarajpoor commented on test.py:
```diff
@@ -1,131 +1,127 @@
+    # test_sdp for cases 1-4
+    modules = utils.import_sdp_mods()
+    for mod in modules:
+        for n_Q in [2, n_T // 2, n_T]:
```
I tried to see how long the coverage testing takes when considering all `n_Q` in `range(2, n_T + 1)` but I noticed it is taking significantly more time (haven't checked but most likely `njit` one is the culprit as we turn off `njit` in coverage testing)  , and I was a bit impatient.  So, I just changed it to `[2, n_T // 2, n_T]`. 

And then I thought: "do we really need to test functions for large `n_T`?" As I think more about this, I am more willing to choose smaller `n_T` for input arrays.

seanlaw replied:
> And then I thought: "do we really need to test functions for large n_T?" As I think more about this, I am more willing to choose smaller n_T for input arrays.

Right. That's why I asked earlier.

Also consider is that you are computing `naive_sliding_dot_product` for every module but you could save time by doing:

```
    for n_Q in [2, n_T // 2, n_T]:
        Q = np.random.rand(n_Q)
        T = np.random.rand(n_T)
        ref = naive_sliding_dot_product(Q, T)
        for mod in modules:
            try:
                comp = mod.sliding_dot_product(Q, T)
                npt.assert_allclose(comp, ref)
            except Exception as e:
                msg = f"Error in {mod.__name__}, with n_Q={n_Q} and n_T={n_T}"
                print(msg)
                raise e
```

NimaSarajpoor replied:
> prime = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, and 97]
and mixed with perfect powers of 2 that are less than 100?

Sounds good. I will add `n_Q == n_T` as well just because it seems like an edge case to me. 


> haven't checked but most likely njit one is the culprit as we turn off njit in coverage testing)
> > It's probably worth checking. It should take the same amount of time as the naive version, right?

Probably not but need to check. The naive one use `np.dot` but the `njit` one uses for-loop, which can become slow when `njit` is off. Will check it.

> Also consider that you are computing naive_sliding_dot_product for every module

Ahh... totally missed it! Right!!!

> Additionally, I wonder.

were you planning to share something else here?

seanlaw replied:
> Additionally, I wonder.

I don't remember now 😄

NimaSarajpoor commented:
@seanlaw 

> I'm not sure I understand what you mean by this.

In STUMPY, for instance, when we create PR and push changes, I can see the tests of GitHub Actions right within that PR. Example: https://github.com/TDAmeritrade/stumpy/pull/1068

I know that the main branch here does not have GitHub Actions yet. ...and I am curious to know if there is a way that this PR can use the GitHub Actions in my branch and shows it right here I this page. I am going through [this discussion](https://github.com/orgs/community/discussions/25746) to see if I can find my answer there.

seanlaw commented:
> I know that the main branch here does not have GitHub Actions yet. ...and I am curious to know if there is a way that this PR can use the GitHub Actions in my branch and shows it right here on this page. I am going through [this discussion](https://github.com/orgs/community/discussions/25746) to see if I can find my answer there.

I would guess that the answer is "no". The workflow is ran from the `main` branch. Once merged then all other PRs should be able to launch the workflow

NimaSarajpoor commented:
>  (and shouldn't because a malicious person could force your repo to run their non-related workflow and that is a security risk)

THIS is one superb perspective!!

NimaSarajpoor commented:
@seanlaw 
The Github-Actions now uses conda to install dependencies and [it is working](https://github.com/NimaSarajpoor/sliding_dot_product/actions/runs/13879092275)

Although the file `environment.yml` includes some extra dependencies, I think we should be okay. What do you think?

----

Btw, we may remove `jax` and `isort` from the file `environment.yml`.

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,65 @@
+  unit-testing:
+    defaults:
+      run:
+        shell: bash -el {0}
```
What exactly is this doing?

seanlaw replied:
Please provide a `name:` to describe it

NimaSarajpoor replied:
> What exactly is this doing?

This is to use login shell (flag `-l` does this here). Otherwise, the conda environment cannot be activated as the non-login bash shell do not use  `~/.profile` or `~/.bashrc` and hence ignores the updates made by `conda init bash` 

Also:
* The flag `-e` is to exit immediately when there is error, and should help with catching error IIUC. 
* `{0}` is a placeholder for script, will be replaced by actual script. (I haven't tried it without `{0}`. I am still trying to better understand it.)


**A few references:**
* [stackoverflow](https://stackoverflow.com/questions/69070754/shell-bash-l-0-in-github-actions)
* [Github Issue for setup-miniconda](https://github.com/conda-incubator/setup-miniconda/issues/128)
* [Important Note in setup-miniconda doc](https://github.com/conda-incubator/setup-miniconda?tab=readme-ov-file#important)


> Please provide a `name:` to describe it

Will do!

seanlaw replied:
Gotcha. I think this note that you showed is good to reference: https://github.com/conda-incubator/setup-miniconda?tab=readme-ov-file#use-a-default-shell or as a comment

seanlaw replied:
Okay, I've done a bit more reading. I think I mostly understand now.

NimaSarajpoor replied:
@seanlaw 
> Something about a login shell vs a non-login shell...
> Okay, I've done a bit more reading. I think I mostly understand now.

Can you please help me understand too? I am trying to put pieces together but I am not sure if I am on the right track.

---

What I've found is that `bash` command in GitHub Actions is equivalent to `bash --noprofile ...` (see [this table](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#defaultsrunshell)) According to [bash manual](https://www.gnu.org/software/bash/manual/bash.html): 

> --noprofile
Don’t load the system-wide startup file /etc/profile or any of the personal initialization files ~/.bash_profile, ~/.bash_login, or ~/.profile when Bash is invoked as a login shell.

And the conda doc says:
> `conda init` will initialize a shell permanently by writing some shell code in the relevant startup scripts of your shell (e.g. ~/.bashrc).

With `shell: bash`, we still get `base` and `test` environments. However, the `conda list` shows installed packages in the path `/Users/runner/miniconda3` which is the path to `base`. 


And this is the confusing part. So, it seems `conda` command can be executed via non-login shell. So, shouldn't it be able to also activate the `test` environment behind the scene?

seanlaw replied:
I should strongly preface my response by admitting that I do not know but here is my best guess (given some level of laziness)! I think there are several factors at play here:

1. Defaults
2. Login shell
3. `conda init`

**Defaults**
So, typically, within a Github Actions workflow, you can run commands from the shell as you're doing here. However, if you look at the [STUMPY workflow](https://github.com/TDAmeritrade/stumpy/blob/main/.github/workflows/github-actions.yml), you'll see that we declare the shell EVERY TIME we execute a shell command. For example

```
        - name: Display Python Version
        run: python -c "import sys; print(sys.version)"
        shell: bash
```

Needing to explicitly add `shell: bash` (note that this is a "non-login shell" because no profile is loaded) every "step" seems tedious and so you can basically set the default shell as you've done here. In the most basic case (i.e., in STUMPY, not here), you could've simply set the default by doing:

```
defaults:
  run:
    shell: bash
```

and then, in the workflow  "steps", you can then omit the repeating of `shell: bash`, which will clean up the workflow (I actually prefer seeing it in every step by to each their own). If you're only using Github Marketplace Actions to do everything and along with `pip` to install packages then this is likely sufficient. 

**Login Shell**
However, since we are using `conda`, which needs to load the `conda` binary locations into our $PATH, then we'll need a "login shell", which will automatically ensure that `conda` can be found in our $PATH. To do this, we switch from `shell: bash` to `shell: bash -l`. It's not a big change but important. Having said that,  don't know why we need the `{0}` except that it is referencing some script itself

**Conda Init**
This is related to the login shell in that `conda init` is often called to setup the shell environment so that `conda` can be found in $PATH. Therefore, this is why we need `shell: bash -l` .

NimaSarajpoor replied:
Thanks for the response and breaking it down to its components!!

I also played a little bit with login-shell vs non-login shell, and checked $PATH. Noticed that `$PATH` contains the path to conda test env only in the login-shell case. There are still some gaps that I need to fill but I think I have most of the pieces of the puzzle.

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,65 @@
+       
+      - name: install pytest and coverage
+        run: |
+          conda install conda-forge::pytest
```
Why not put these inside of your `environment.yml`?

NimaSarajpoor replied:
To be consistent with [the environment.yml file in stumpy](https://github.com/TDAmeritrade/stumpy/blob/main/environment.yml), I will add those two libraries to `environment.yml `.

----

I wasn't sure about the **scope** of the existing `environment.yml`, available in the `main` branch. It seems that it contains all the libraries that are required for running the program and developing it. So, on one hand, I thought `pytest` and `coverage` are only for testing, and maybe should be kept outside of `environment.yml`.

seanlaw replied:
I think that since it isn't being used to build a package that we should be okay

seanlaw commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+numpy>=1.22
+scipy>=1.10
+numba>=0.59.1
+pyFFTW>=0.15.0
```
Can we remove this file  completely now?

NimaSarajpoor replied:
Yes, and I think I should because it does not reflect the FFTW dependency.

seanlaw commented on sdp/challenger_sdp.py:
```diff
@@ -1,7 +1,7 @@
 
 
-def setup(Q, T):
+def setup(Q, T):  # pragma: no cover
```
I think we should test that each module has this function and that it returns `None`. Doing so would ensure that this function always exists.

NimaSarajpoor replied:
Good point! Will do!

seanlaw reviewed (changes requested):
@NimaSarajpoor I left some comments. I don't see any `environment.yml` file though

NimaSarajpoor commented:
@seanlaw 

> I don't see any environment.yml file though (maybe you forgot to commit it?)

I am just using the one that is available in the `main` branch. I will modify it according to the recent comments.
-

NimaSarajpoor commented:
@seanlaw 
I've addressed most of the comments. 

(1) Changed `branches: '**'` to `branches: add_test` for now so that I can still see GitHub Actions in my fork. Just before merge, will change it to `branches: main`
(2) Couldn't figure out how to add `name:` to the part where we set default `bash` to `shell: bash -el {0}`. Added some comments only.

NimaSarajpoor commented on timing.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+./test.sh
```
As suggested, now executing shell script rather than sourcing it. 

**[Note to self]**
To understand executing-vs-sourcing, see [this stackoverflow](https://superuser.com/questions/176783/what-is-the-difference-between-executing-a-bash-script-vs-sourcing-it)

----

@seanlaw 
we can try to just do unit testing part here but having coverage should be okay given the fact that the running time is small.

seanlaw replied:
I'm not sure. I was thinking that we DON'T do testing here and only do timing here and keep testing separate? 

Having said that, we can follow STUMPY and do both unit testing and coverage testing together inside of `test.sh`.

NimaSarajpoor replied:
Cool! Will separate them! `test.sh` 

Currently, `test.sh` follows STUMPY's design, and it does both unit testing and coverage.

NimaSarajpoor commented on test.py:
```diff
@@ -1,131 +1,177 @@
+    Q = np.random.rand(3)
+    T = np.random.rand(10)
+
+    modules = utils.import_sdp_mods()
```
@NimaSarajpoor 
This test does not make sense because `utils.import_sdp_mods()` checks modules and only import them if they have the function `setup`.  So, the actual test we are doing here is only for the returned value.  We want to check ALL modules in the `spd/` sand make sure they ALL have the function `setup`. 

**Next step:** 
Revise the test function by checking all modules in `sdp` and make sure it has both functions `sliding_dot_product` and `setup`. (Its downside is that we cannot easily create and use any helper module in sdp for `challenger_sdp`)

NimaSarajpoor reviewed (commented):
@seanlaw 
I also added a few comments for myself. Can you please look at them and let me know what you think?

seanlaw commented on timing.py:
```diff
@@ -0,0 +1,105 @@
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="./test.py -noheader -pmin 6 -pmax 23 -pdiff 3 pyfftw challenger"
```
This should say `./timing.py`mj0w

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,63 @@
+        uses: actions/checkout@v4
+
+      - name: Update environment.yml with PYTHON_VERSION
+        run: sed 's/python.*/python==${{ env.PYTHON_VERSION }}/' environment.yml > environment_new.yml
```
I think that your regex is too loose. I would be more specific with:

```
sed -r 's/- python[>=]+[0-9]+\.[0-9]+/- python==${{ env.PYTHON_VERSION }}/
```

1. `-r` uses "regex extended" as `sed` by default uses "basic regular expression" (BRE), which is very limited
2. match on `- python` rather than just `python` for specificity
3. `[>=]+` matches "one or more characters found in the set '(>=)'"
4. `[0-9]+` matches "one or more digits"

NimaSarajpoor replied:
> your regex is too loose

Agree. I was just searching for ANY python. IIUC, your concern is for a scenario where this can go wrong. For instance, if another package is added that has `python` as part of its name.  Thanks for the note 👍  

> -r uses "regex extended" as sed by default uses "basic regular expression" (BRE),

Good to know 👍

seanlaw replied:
> IIUC, your concern is for a scenario where this can go wrong. For instance, if another package is added that has python as part of its name.

Yes, exactly!

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,63 @@
+      - name: Display environment_new.yml
+        run: cat environment_new.yml
+       
+      - name: set up conda and install dependencies
```
1. replace `conda` with `miniconda`
2. "install new dependencies" (to reference the new environment.yml)

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,63 @@
+      - name : Remove environment_new.yml
+        run: rm -rf environment_new.yml
+       
+      - name : Display Conda Environments
```
Please choose a consistent naming convention (i.e., title case)

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,63 @@
+      - name : Display Conda Environments
+        run: conda env list
+       
+      - name: Display conda info
```
Is this necessary/helpful?

NimaSarajpoor replied:
Not really. The only helpful information is `active environment`.However, I can get that information 
from the step  `Display Conda Envionrmen`. So, going to remove this.

seanlaw commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,63 @@
+      - name: Display conda info
+        run: conda info
+       
+      - name: Display conda list
```
"Display all installed packages"

seanlaw commented on test.py:
```diff
@@ -1,131 +1,188 @@
+# 1000-2000. This allows us to test sliding_dot_product for
+# wider range of query lengths.
+
+test_inputs = [
```
Maybe to make it easier for our future self, we can remind ourselves what the three values are in a comment:

```
test_inputs = [
    # Input format:
    # (
    #     len(T),
    #     remainder,  #  from `len(T) % 2`  
    #     comparator,  # for len(T) comparator next_fast_len(len(T))
    # )
    (
        2 * (3**2) * (5**3),
        0,
        eq,
    ),  # = 2250, Even `len(T)`, and `len(T) == next_fast_len(len(T), real=True)`
]
```

NimaSarajpoor replied:
very cool! Sure!!

NimaSarajpoor commented on .github/workflows/github-actions.yml:
```diff
@@ -0,0 +1,63 @@
+name: Tests
+on:
+  push:
+    branches: add_test
```
Right. The only thing I noticed is that if I switch to `main` NOW, the Github Actions in my fork will not be triggered. So, I will change it at the end just before merge if that's okay. I will test this behaviour again and if I notice having `main` can result in triggering Github Actions, I will replace it with `main` before resolving this conversation.  I will report back.

Sounds good?

NimaSarajpoor replied:
>  I will report back.

Checked again... Having `main` results in not triggering Github-Actions in my fork's branch, which makes sense. Having said that, I've decided to change it to `main` NOW so that we do not need to keep this task in mind.

And, whenever I need to check GitHub Actions, I will just replace `main` with `MY_OTHER_BRANCH` in MY_OTHER_BRANCH so that I do not create confusion in this PR.

NimaSarajpoor commented on test.py:
```diff
@@ -1,131 +1,194 @@
-            and func_exists(mod_path, "sliding_dot_product")
-            and func_exists(mod_path, "setup")
-        ):
+        if m[1].endswith("_sdp"):
```
@NimaSarajpoor 
The test function now checks all modules that end with `_sdp`. To make it consistent with what we have in `utils.import_sdp_mods`, we should change this condition to: `if sdp in m[1]`.

seanlaw commented:
> Btw, regarding long running time of coverage that we faced before. I was supposed to check the performance of njit-vs-naive when njit is off. I did and noticed that njit is considerably slower. Going to attach the timing here just in case.
[timing.csv](https://github.com/user-attachments/files/19354959/timing.csv)

What does this mean for coverage testing though? How long are the unit tests and coverage tests taking now?

seanlaw commented:
Thanks for working on this @NimaSarajpoor!

NimaSarajpoor commented:
@seanlaw 
> What does this mean for coverage testing though? 

we are good now as the current tests are only for a small number of queries, with short length. I just did not want to leave that question unanswered.

How long are the unit tests and coverage tests taking now?

The unit test and coverage step takes less than 1min in each platform.
